### PR TITLE
add Minecraft Server Script

### PR DESCRIPTION
**Pull request by theofficialgman**
~~currently in draft form, this is not finalized (or tested) yet~~

closes https://github.com/Botspot/pi-apps/issues/1344 closes https://github.com/Botspot/pi-apps/issues/1250 closes https://github.com/Botspot/pi-apps/issues/1304

### CODE REVIEW

**Comment by theofficialgman** [comment]
@GYKGamer I hope the app/script is up to your approval
much of it has changed since your original zip

**Comment by GYKgamer** [comment]
Did you take that from my script?

**Comment by theofficialgman** [comment]
> Did you take that from my script?

I started with your script as a base. You can see that in the first commit in this PR. Almost all of the code has changed since to better perform what was desired

**Comment by theofficialgman** [comment]
I'm marking this script as ready for review/merging
@GYKgamer @Crilum @Botspot please take your final looks/testing of it

File: apps/Minecraft Java Server/install [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+Icon=${DIRECTORY}/apps/Minecraft Java Server/icon-64.png
+Type=Application
+Categories=Game
+Terminal=true
+EOF
+
+
+status "Your IP address is one of the following: 
+$(hostname -I | grep -Eo '[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}')"
+status "The IP address is used to play on your local network"
```
Comment (Crilum) [inline apps/Minecraft Java Server/install:189]:
```suggestion
status "The IP address is used to play on your local network."
```

I'm just that kind of person who likes grammar.. :+1:

File: apps/Minecraft Java Server/uninstall [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+
+shopt -s extglob
+
+#skip uninstallation if executed from an update
+if [[ "$1" != "update" ]];then
```
Comment (Crilum) [inline apps/Minecraft Java Server/uninstall:6]: Could this break updates?

**Review by Crilum — COMMENTED** [review]
I think it's good. One grammar issue (missing period), and one question. Will installing the server again, without removing it, break things?

File: apps/Minecraft Java Server/install (hunk 1 above)
Comment (theofficialgman) [reply apps/Minecraft Java Server/install]: added

**Comment by Crilum** [comment]
Yeah, ok. :+1:

**Comment by GYKgamer** [comment]
Tested it, works fine. I'm guessing should be finished soon?

**Comment by theofficialgman** [comment]
yes, just need to copy over the "more verbose changes" that bospot wanted

**Comment by GYKgamer** [comment]
less go.
An idea i got nearly a month ago is in a working phase

**Comment by theofficialgman** [comment]
force pushed to rebase on master branch and cleanup history
no changes made

File: apps/Minecraft Java Server/install [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                if [ $? != 0 ];then
+                    sudo rm -f /etc/apt/sources.list.d/adoptium.list /usr/share/keyrings/adoptium-archive-keyring.gpg
+                    error "Failed to install temurin packages"
+                fi
+                java_location="/usr/lib/jvm/java-17-temurin-armhf/bin/java"
+                ;;
+        esac
+        ;;
+    *)
+        error "arch variable is not set, can not continue"
```
Comment (derryktutt) [inline apps/Minecraft Java Server/install:125]: error "arch variable is not set, can not continue" should be
  error "Arch variable is not set, can not continue".

**Review by derryktutt — CHANGES_REQUESTED** [review]

File: apps/Minecraft Java Server/install (hunk 3 above)
Comment (theofficialgman) [reply apps/Minecraft Java Server/install:125]: I left it lowercase because the variable name is lowercase

**Comment by GYKgamer** [comment]
Downloadable version out yet?

**Comment by theofficialgman** [comment]
this hasn't been merged, but its done

**Comment by GYKgamer** [comment]
alr

File: apps/Minecraft Java Server/install (hunk 3 above)
Comment (Crilum) [reply apps/Minecraft Java Server/install:125]: We could add a period, though..

**Comment by theofficialgman** [comment]
I've noticed this has now been requested twice now:
https://github.com/Botspot/pi-apps/issues/1344 https://github.com/Botspot/pi-apps/issues/1250 

however, I'm not sure if I like the implementation for starting/stopping the minecraft server as it is. I've recently started a server myself (that I host) and found this script to not be suitable for setting it up

using a systemd service for starting/stopping the server seems much more practical. I've implemented this myself but have not made changes here yet

@GYKgamer I'd like your input. 

would you rather have systemd handle starting/stopping the server (like this `sudo systemctl start minecraft` `sudo systemctl stop minecraft`) which opens the possibility for running it on startup, as well as doing safe shutdowns on restart/shutodown of the system? 
if its started, you can attach to the currently running server with `screen -r Minecraft_Server` and detach as with any other screen command `ctrl+A` then `D`

**Comment by GYKgamer** [comment]
> I've noticed this has now been requested twice now: #1344 #1250
> 
> however, I'm not sure if I like the implementation for starting/stopping the minecraft server as it is. I've recently started a server myself (that I host) and found this script to not be suitable for setting it up
> 
> using a systemd service for starting/stopping the server seems much more practical. I've implemented this myself but have not made changes here yet
> 
> @GYKgamer I'd like your input.
> 
> would you rather have systemd handle starting/stopping the server (like this `sudo systemctl start minecraft` `sudo systemctl stop minecraft`) which opens the possibility for running it on startup, as well as doing safe shutdowns on restart/shutodown of the system? if its started, you can attach to the currently running server with `screen -r Minecraft_Server` and detach as with any other screen command `ctrl+A` then `D`
> 
> I also realize that some users may rather use forge/fabric than vanilla minecraft, so that should be added to this script as well

The systemd handle seems like a good idea, although normal users would have a hard time trying to remeber the command (unless we tell them the command) and the forge/fabric idea is a good idea as i've thought of it before but it'd be hard to try and implement it in but is possible. In the topic of different minecraft versions I found this website https://mcversions.net/ and they could input a version in and them be able to put different version of the server in.

**Comment by GYKgamer** [comment]
> > I've noticed this has now been requested twice now: #1344 #1250
> > however, I'm not sure if I like the implementation for starting/stopping the minecraft server as it is. I've recently started a server myself (that I host) and found this script to not be suitable for setting it up
> > using a systemd service for starting/stopping the server seems much more practical. I've implemented this myself but have not made changes here yet
> > @GYKgamer I'd like your input.
> > would you rather have systemd handle starting/stopping the server (like this `sudo systemctl start minecraft` `sudo systemctl stop minecraft`) which opens the possibility for running it on startup, as well as doing safe shutdowns on restart/shutodown of the system? if its started, you can attach to the currently running server with `screen -r Minecraft_Server` and detach as with any other screen command `ctrl+A` then `D`
> > I also realize that some users may rather use forge/fabric than vanilla minecraft, so that should be added to this script as well
> 
> The systemd handle seems like a good idea, although normal users would have a hard time trying to remeber the command (unless we tell them the command) and the forge/fabric idea is a good idea as i've thought of it before but it'd be hard to try and implement it in but is possible. In the topic of different minecraft versions I found this website https://mcversions.net/ and they (person setting up the server) could input a version in and them be able to put different version of the server in although it does not sadly have forge, fabric etc.

**Comment by GYKgamer** [comment]
ah well then in that case my expectations were wrong and it is actually easy to code the websites in. Thanks for informing me

**Comment by theofficialgman** [comment]
@GYKgamer alright all that is done
I'd appreciate you trying it out

**Comment by GYKgamer** [comment]
can't do this today or possibly tommorow but will try asap

**Comment by theofficialgman** [comment]
@GYKgamer @Crilum @Botspot I'm once again to the point that I am happy with this script

maybe in the future I'll make a PR to add CraftBukkit, Spigot, and/or Paper, but for now this should do

**Comment by rbcoder117** [comment]
@theofficialgman is this tread active still or is it dead?

**Comment by theofficialgman** [comment]
@rbcoder117 this PR has been ready to merge, don't know why it hasn't been yet

To close: https://github.com/Botspot/pi-apps/issues/1344 https://github.com/Botspot/pi-apps/issues/1250 https://github.com/Botspot/pi-apps/issues/1304

**Comment by rbcoder117** [comment]
ok that is what I thought.
@Botspot is this ready to merge?

**Comment by Botspot** [comment]
> ok that is what I thought. @Botspot is this ready to merge?

It looks pretty close. I just tested the script and it completed successfully.
I wanted to make a few minor improvements to the script, but don't seem to be able to. @theofficialgman, is there a permission you need to change to allow maintainers to make changes to the PR?

**Comment by theofficialgman** [comment]
> It looks pretty close. I just tested the script and it completed successfully. I wanted to make a few minor improvements to the script, but don't seem to be able to. @theofficialgman, is there a permission you need to change to allow maintainers to make changes to the PR?

yes I usually disable the "allow maintainers to make change to the PR" because I have a GitHub personal access token in my repo. If I had github actions enabled on my repo (which I sometimes do when I'm testing stuff), you could potentially push malicious changes to this PR (which would go into my repo) triggering a github actions which would give you my PAT. Access to that PAT would affectively allow you to push to my repo without my consent unless I deleted the token.

Anyway, I don't need that PAT in my repo anymore (it was for testing the wiki changes), so I can remove it and you to make changes in a second

**Comment by theofficialgman** [comment]
@Botspot you should be able to make changes now

**Comment by Botspot** [comment]
Thanks everyone for the great work!

**Comment by rbcoder117** [comment]
YaY

**Comment by GYKgamer** [comment]
Well done everyone

**Comment by Botspot** [comment]
Apps like this really show what a community can do, when working together. I'm not sure one person could have made this app on their own without other's help. Think about it - this app required minecraft knowledge, java knowledge, bash knowledge, yad knowledge, english knowledge, etc.
Each of us have our strengths, and when we put them together, wonderful things can happen.

**Comment by GYKgamer** [comment]
the only reason why i started this was because I was wanting the normal server installer that all guides used for this lol. Now it became a selectable choice that people will a) find easier b) much more better then original methods as they just used the latest mc version

**Comment by theofficialgman** [comment]
I just want to post that I've been using this yad function for about 6 months now on the L4T-Megascript
I called it `userinput_func`

```
function userinput_func {
  unset uniq_selection
  height=$(($(echo "$description" | grep -o '\\n' | wc -l) + 8))
  height_gui=$(echo $(( height * 15 + ${#@} * 20 + 100 )))
  if [[ $gui == "gui" ]]; then
    if [[ "${#@}" == "2" ]];then
      echo -e "$1" |  yad --show-uri --center --image "dialog-information" --borders="20" --title "User Info Prompt" \
      --text-info --fontname="@font@ 11" --wrap --width=800 --height=500 \
      --show-uri  --window-icon=/usr/share/icons/L4T-Megascript.png \
      --button="$2":0
      output="$2"
    elif [[ "${#@}" == "3" ]];then
      yad --image "dialog-question" \
      --borders="20" --height="200" --center --fixed\
      --window-icon=/usr/share/icons/L4T-Megascript.png \
      --text="$1" \
      --button="$2":0 \
      --button="$3":1
      if [[ $? -ne 0 ]]; then
        output="$3"
      else
        output="$2"
      fi
    else
      for string in "${@:2}"; do
        uniq_selection+=(FALSE "$string")
      done
      uniq_selection[0]=TRUE
      output=$(
        yad --center --fixed --height=$height_gui\
          --borders="20" \
          --window-icon=/usr/share/icons/L4T-Megascript.png \
          --text "$1" \
          --list \
          --radiolist \
          --column "" \
          --column "Selection" \
          --print-column=2 \
          --separator='' \
          --button="Ok":0 \
          "${uniq_selection[@]}"
      )
    fi
  else
    if [[ "${#@}" == "3" ]];then
      dialog  --stdout --clear --yes-label "$2" --no-label "$3" --yesno "$1" "$height" "120"
      if [[ $? -ne 0 ]]; then
        output="$3"
      else
        output="$2"
      fi
    else
      for string in "${@:2}"; do
        uniq_selection+=("$string" "")
      done
      output=$(
        dialog --stdout --clear \
          --backtitle "CLI Chooser Helper" \
          --title "Choices" \
          --menu "$1" \
          "$height" "120" "$($# - 1)" \
          "${uniq_selection[@]}"
      )
    fi
  fi
  status=$?
  if [[ $status = "1" ]]; then
    clear -x
    echo "Canceling Install, Back to the Megascript"
    exit 1
  fi
  clear -x
}
```
its what the yad function in this script is derived from
it takes a couple of input variables, and automatically creates a yad prompt with "decent" sizing, button or list format depending on the number of selections, and also will create a dialog box instead of yad for CLI only use
(set `gui=gui` for yad, and `gui=cli` for dialog)

@Botspot you are welcome to include a form of this function in the pi-apps api if you would like, I find it handy to get all sorts of user input

**Comment by GYKgamer** [comment]
Hi, sorry for not being able to test any software, but I finally found my rpi 4 :). I can test it now

**Comment by theofficialgman** [comment]
> Hi, sorry for not being able to test any software, but I finally found my rpi 4 :). I can test it now

@GYKgamer no worries, but this has long been merged into pi-apps and made even better, don't try this old PR

just use it if you would like, everything should work nicely

**Pull request merged**